Using array object as data store

XedinUnknown opened the pull request:
Aims to incorporate solution for #4.

XedinUnknown committed:
Added autoload-dev paths to tests

XedinUnknown committed:
Added test for `NormalizeKeyCapableTrait`

XedinUnknown committed:
Added `NormalizeKeyCapableTrait`

XedinUnknown committed:
Some traits now use the new `_normalizeKey()` method

XedinUnknown committed:
`GetDataStoreCapableTrait` must now return an `ArrayObject`

XedinUnknown committed:
`GetDataStoreCapableTrait` now returns an `ArrayObject`

XedinUnknown committed:
`GetDataCapableTrait` now expected to work with `ArrayAccess`

XedinUnknown committed:
`GetDataCapableTrait` now works with array access and requires key

It returns only one piece of data always.

XedinUnknown committed:
`HasDataCapableTrait` now expected to work with `ArrayAccess`

XedinUnknown committed:
Fixed documented class name not being imported

XedinUnknown committed:
`HasDataCapableTrait` now works with `ArrayAccess`

XedinUnknown committed:
`UnsetDataCapableInterface` has its expectations changed

- Now must work with `ArrayAccess`.
- Now must unset a single key.

XedinUnknown committed:
Unset data behaviour changed

- Now works with `ArrayAccess`.
- Now unsets a single key.

XedinUnknown committed:
Added test for `UnsetManyCapableTrait`

XedinUnknown committed:
Added `UnsetManyCapableTrait`

XedinUnknown committed:
CS fixed

XedinUnknown committed:
Added test for `SetManyCapableTrait`

XedinUnknown committed:
Added `SetManyCapableTrait`

XedinUnknown committed:
Changed expectations for `SetDataCapableTrait`

- Must now work with `ArrayAccess`.
- Must now set a single key.

XedinUnknown committed:
Changed behaviour of `SetDataCapableTrait`

- Now works with `ArrayAccess`.
- Now sets a single piece of data.

XedinUnknown committed:
Removed `AbstractDataObject`

Has no additional functionality, but only assembles trait.
Looks redundant.

XedinUnknown committed:
Added PHPS project files

mecha commented on src/GetDataStoreCapableTrait.php:
```diff
@@ -23,12 +25,12 @@
     {
         return $this->dataStore === null
-                ? $this->dataStore = new \stdClass()
+                ? $this->dataStore = new ArrayObject()
```
Why not abstract this into a `_createDataStore() : ArrayAccess` method? It would certainly be a beneficial abstraction to allow any `ArrayAccess` implementation. A `CreateDataStoreCapableTrait` can complement this.

mecha replied:
If you agree, be sure to change the `ArrayObject` types in the docs to `ArrayAccess`.

XedinUnknown replied:
Perhaps the abstraction could be useful. But let's talk about the hinted type.
For the implementations in this package, `ArrayAccess` is all that's necessary. However, a part of the reason for this whole changeset is the `MapInterface` implementation (and its countable version). The first requires the storage to in addition be `iterable`, and the second - `Countable`. What do we type-hint and where?

mecha replied:
I would still type hint `ArrayAccess` in this package. Like you said, it's all that's necessary here.

I anticipate that the map abstract code (be it a class or trait) will depend on some data store via an abstract method. _That_ method can document that it requires an iterable. The same applies for whatever abstract code will do the counting.

When the concrete or base classes are being composed, it is then the developer's job to make sure that everything is compatible. As long as the necessary abstraction is provided. A few checks also wouldn't hurt.

```php
class CountingAndStuff implements Countable
{
	// This class is now completely aware that the data store is an `ArrayObject`
	use CreateDataStoreCapableArrayObjectTrait;

    public function count()
    {
        $dataStore = $this->_getDataStore();

		// Still doesn't hurt to check, for completeness' sake. I actually encourage this anyway
		if ($dataStore instanceof Countable) {
			return $dataStore->count();
		}

		// Even if this might never happen, it's a good safeguard against potential instance setup problems
		throw new InvalidArgumentException('Data store is not countable.');
    }
}
```

Sure, that might not look anything like the final code, but it helps get my point across. Everything remains concerned with its own things. The alternative is to create a `DataStoreInterface`. But even then we'd probably end up segregating the interfaces and end up with the same problem and same solution.

mecha commented on src/NormalizeKeyCapableTrait.php:
```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: xedin
```
Well hello there!

mecha commented on src/SetManyCapableTrait.php:
```diff
@@ -0,0 +1,110 @@
+            try {
+                $this->_setData($_key, $_value);
+            } catch (InvalidArgumentException $e) {
+                throw $this->_createOutOfRangeException(
```
Surely this must fit in 1 line. Can't it? 🤣

XedinUnknown replied:
Prolly not

mecha replied:
I actually checked before writing this comment. It does 😛

XedinUnknown replied:
Anything can fit in one line. But should it?

mecha replied:
Oh crap. You're right. Should it? Ok, you have my go ahead. Fit everything into 1 line.

mecha replied:
I should probably follow up on that. That was a joke. Please don't obfuscate ✌️ 

I only brought this up because I like clean code. It doesn't bother me _that_ much.

XedinUnknown replied:
It was you who started writing these things in the multi-line fashion. Since I'm all about standards, I decided that all exception factory methods that specify the 4th parameter should be written this way, regardless of length.

mecha commented on src/UnsetManyCapableTrait.php:
```diff
@@ -0,0 +1,127 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: xedin
```
He be `xedin` light on da wae!

mecha commented on src/UnsetManyCapableTrait.php:
```diff
@@ -0,0 +1,127 @@
+
+        foreach ($keys as $_idx => $_key) {
+            try {
+                $this->_unsetData($_key);
```
Abstract method `_unsetData()` is not declared in this trait.

XedinUnknown replied:
Good catch.

mecha reviewed (changes requested):
I must say, I knew using `ArrayAccess` as the data store would make these traits look cleaner, but this is just beautiful. Just a few things to fix here and there.

XedinUnknown committed:
Removed redundand file docs

XedinUnknown committed:
Cleaned up trait

- Added missing abstract method.
- Removed redundant abstract methods.

This fixes #4.

XedinUnknown committed:
Added test for `SetDataStoreCapableTrait`

XedinUnknown committed:
Added `SetDataStoreCapableTrait`

XedinUnknown committed:
Added test for `CreateDataStoreCapableTrait`

XedinUnknown committed:
Added `CreateDataStoreCapableTrait`

XedinUnknown committed:
Changed expectations for `GetDataStoreCapableTrait`

Now must use store factory

XedinUnknown committed:
`GetDataStoreCapableTrait` now uses data store factory

XedinUnknown committed:
Combined data store setter and getter traits into one aware trait

XedinUnknown commented:
Commit 5023d5b allows us to circumvent [bug 61943][php-bug\61943], which causes notices on versions of PHP prior to PHP 7.0.

[php-bug\61943]: https://bugs.php.net/bug.php?id=61943

mecha reviewed (approved):
LGTM. The merging of the getter and setter traits took me by surprise, but after thinking about it for a while, it makes more sense and looks more consistent with other Dhii traits.
